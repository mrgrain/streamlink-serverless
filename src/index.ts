import { join } from 'path';
import * as cdk from 'aws-cdk-lib';
import { BundlingFileAccess, BundlingOutput, DockerImage } from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

export interface StreamlinkProps {
}

export class Streamlink extends Construct {
  public readonly function: lambda.Function;

  public constructor(scope: Construct, id: string, _props: StreamlinkProps = {}) {
    super(scope, id);

    const runtime = Runtime.PYTHON_3_9;
    const streamlinkApp = join(__dirname, '..', 'streamlink');

    const layer = new lambda.LayerVersion(this, 'Layer', {
      compatibleRuntimes: [runtime],
      code: lambda.Code.fromAsset(streamlinkApp, {
        bundling: {
          image: DockerImage.fromBuild(streamlinkApp),
          workingDirectory: '/build',
          command: [
            'bash', '-c',
            `pip install --upgrade aws_lambda_types streamlink lxml -t python/lib/${runtime.name}/site-packages/ && zip -rq /asset-output/layer.zip python`,
          ],
          outputType: BundlingOutput.ARCHIVED,
          bundlingFileAccess: BundlingFileAccess.VOLUME_COPY,
        },
      }),
    });

    this.function = new lambda.Function(this, 'Default', {
      runtime,
      handler: 'app.handler',
      code: lambda.Code.fromAsset(streamlinkApp),
      memorySize: 512,
      timeout: cdk.Duration.seconds(5),
      layers: [layer],
    });
  }
}