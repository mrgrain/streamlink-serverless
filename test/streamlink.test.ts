import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Streamlink } from '../src';

describe('Streamlink', () => {
  test('creates a Function', () => {
    const stack = new cdk.Stack();
    new Streamlink(stack, 'MyStreamlink', {});

    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });
});