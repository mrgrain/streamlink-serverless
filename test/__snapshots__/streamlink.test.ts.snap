// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Streamlink creates a Function 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "MyStreamlink84573151": Object {
      "DependsOn": Array [
        "MyStreamlinkServiceRole0284652C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "1d3a0f69575846562ba1d82d5ffe1e2a0bdb43c24151aecd20072ee35bff47c6.zip",
        },
        "Handler": "app.handler",
        "Layers": Array [
          Object {
            "Ref": "MyStreamlinkLayer5A15AFB8",
          },
        ],
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyStreamlinkServiceRole0284652C",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyStreamlinkLayer5A15AFB8": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "python3.9",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f6191b25ca59c148725bb21616495d59f660596ba6e24a6261d6adc5144078ba.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "MyStreamlinkServiceRole0284652C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
